### PR TITLE
Client Regex Blocking

### PR DESCRIPTION
**Pull request by LewdNeko-chan**
So, I decided to add regex blocking, because I don't want some bad clients (known CVEs and so on, on my tracker), like you see in this file.
So, I added some functions, to block clients with the regex function.

Best Regards,
neko.

### CODE REVIEW

**Review by leighmacdonald — CHANGES_REQUESTED** [review]
Looks alright, but has a few issues code wise that need fixing. 

Beyond that i have other concerns though:

The performance characteristic of regex are not great, especially when in a loop. There is a high cost of creating the initial compiled regex object the first time it get called vs something basic like string comparison. Subsequent calls for the same regex pattern will use the compiled version at least. But this will scale worse the more clients that get blacklisted.

What is the advantage over something much simpler like this using the first 8 bytes of the PeerID, which is basically what actual torrent clients do to show you the peer clients.

`if (!in_array(get_client_prefix(peer_id), ["your", "client", "list"])) { return false; }`

An example in go https://github.com/leighmacdonald/mika/blob/master/store/peer.go#L266

File: app/Http/Controllers/AnnounceController.php [hunk 1]
```diff
@@ -166,8 +166,11 @@ protected function checkClient(Request $request): void
         }
 
         // Block Blacklisted Clients
-        if (in_array($request->header('User-Agent'), config('client-blacklist.clients'))) {
-            throw new TrackerException(128, [':ua' => $request->header('User-Agent')]);
+        foreach (config('client-blacklist.clients') as $value) {
+            if(preg_match($value, "".$request->header('User-Agent').""))
```
[the remark this answers is not in the document]
Comment (LewdNeko-chan) [reply app/Http/Controllers/AnnounceController.php:170]: Didn't know, should have removed that, hat that for debugging, forget to remove.

File: config/client-blacklist.php [hunk 2]
```diff
@@ -31,9 +31,32 @@
...
+
+    /*
+    |--------------------------------------------------------------------------
+    | Blacklist Clients
+    |--------------------------------------------------------------------------
+	|
+	| Don't forget to add the readable on this array for display on the page.
+	|
+	*/
+	'clientsreadable' => [
```
[the remark this answers is not in the document]
Comment (LewdNeko-chan) [reply config/client-blacklist.php:58]: Seems legit.

**Comment by LewdNeko-chan** [comment]
> Subsequent calls for the same regex pattern will use the compiled version at least. But this will scale worse the more clients that get blacklisted.

Then is there maybe still a possibility to only use regex or the old method on the webpage?

**Comment by Alkl58** [comment]
> What is the advantage over something much simpler like this using the first 8 bytes of the PeerID, which is basically what actual torrent clients do to show you the peer clients.
> 
> `if (!in_array(get_client_prefix(peer_id), ["your", "client", "list"])) { return false; }`
> 
> An example in go https://github.com/leighmacdonald/mika/blob/master/store/peer.go#L266

With regex you are free to choose which versions of speicifc clients get's blacklisted.
It would be a hastle to write every µTorrent version in that list.
Your in go written example has the problem that the first 8 bytes of the peer id is not enough to block specific versions of for example qBittorrent. (I might be wrong, but thats what I understand)

**Comment by leighmacdonald** [comment]
The versions get extracted from the peerid for qb, see https://github.com/qbittorrent/qBittorrent/wiki/Frequently-Asked-Questions#What_is_qBittorrent_Peer_ID. There just needs to be a bit of extra logic to ensure is within the allowed major/minor/patch levels.

**Pull request closed without merging**